feat(BUX-212/237: CODEOWNERS, CODE_STANDARDS (#1)
4chain-ag/code-standards-bsv: .github/CODE_STANDARDS.md

# Code Standards & Contributing Guidelines

## Table of Contents

- [Code Standards \& Contributing Guidelines](#code-standards--contributing-guidelines)

>   - [Table of Contents](#table-of-contents)

Reviewer: Unnecessary link

Author: You are right - it was automatically created by my VSCode extension, sorry.

      - [As a PR Reviewer](#as-a-pr-reviewer)
    - [2.2. Tests](#22-tests)
      - [Principle](#principle)
      - [Guidelines for Writing Tests](#guidelines-for-writing-tests)
        - [Pull request title with a scope and task number](#pull-request-title-with-a-scope-and-task-number)
    - [3.3 Branching](#33-branching)
      - [Choosing branch names](#choosing-branch-names)
      - [Descriptiveness](#descriptiveness)
      - [Include Issue Number](#include-issue-number)
      - [Deleting Branches After Merging](#deleting-branches-after-merging)
      - [Remove Remote Branches](#remove-remote-branches)
      - [Recommendation: Clean Local Branches](#recommendation-clean-local-branches)
  - [4. Documentation Code Standards](#4-documentation-code-standards)
    - [4.1 Overview](#41-overview)
    - [4.2 Principles](#42-principles)
    - [4.3 Feature Documentation](#43-feature-documentation)
      - [4.3.1 Necessity](#431-necessity)
      - [4.3.2 Examples](#432-examples)
    - [4.4 External Features](#44-external-features)
    - [4.5 Markdown usage](#45-markdown-usage)
    - [4.5 Conclusion](#45-conclusion)

## Most important rules - Quick Checklist

> - [ ] Follow official Go guidelines for style and formatting.

Reviewer: Checklist? Bullet list or smth wouldn't suit better?

Reviewer: This one is only question/suggestion not a blocker ;-)

- [ ] Write self-documenting code and minimize comments.
- [ ] Ensure comprehensive test coverage including happy and error paths.
- [ ] Provide meaningful and constructive code reviews.
- [ ] Adhere to Conventional Commits for commit messages and PR naming.
- [ ] Document every feature adequately, especially for open-source projects.
- [ ] Keep documentation clear, concise, up-to-date, and accessible.
- [ ] Branching - choose consistent naming conventions, include issue number, delete branches after merging.

## 1. Code style and formatting - official guidelines

### 1.1 In GO applications or libraries, we follow the official guidelines

- [Effective Go](https://go.dev/doc/effective_go) - official Go guidelines

> - [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments) - official Go code review comments

Reviewer: What a great list!

- [Go Test](https://pkg.go.dev/testing) - official Go testing package & recommendations
- [Go Linter](https://golangci-lint.run/) - golangci-lint - only codestyle checks

 > Our current linter configuration is in the `.golangci.yml` file.

#### Additional useful resources with GO recommendations, best practices and the common mistakes

- [Go Styles by Google](https://google.github.io/styleguide/go/) - Google's Go Style Guide
- [Uber Go Style Guide](https://github.com/uber-go/guide/blob/master/style.md) - Uber's Go Style Guide
- [Go Common Mistakes](https://github.com/golang/go/wiki/CommonMistakes) - Common Mistakes in Go

## 2. Code Rules

### 2.1. Self-documenting code

#### As a Developer

- Refactoring tasks should be strategically undertaken; not every piece of code warrants modification. If a segment of code, despite comments, remains untouched and unproblematic, it is   likely fulfilling its purpose effectively.
- When embarking on feature additions or bug fixes, and encountering commented code, consider extracting functions and creating a PR before proceeding with the primary task. This helps maintain code clarity and function.

> - Should you come across a superfluous comment during your changes, do take the initiative to remove it and create a PR for that. This practice contributes to keeping the repository neat and well-maintained.

Reviewer: I'm not sure if the sentence is written correctly
```suggestion
- If you come across a superfluous comment during your changes, take the initiative to remove it and create a PR for that. This practice contributes to keeping the repository neat and well-maintained.
```

Reviewer 2: It is correct, a bit more formal and sublime tho :)

#### Guidelines for Code Review

1. **Constructive Feedback**: Reviewers should provide constructive feedback, highlighting both strengths and areas of improvement. Comments should be clear, concise, and related to code structure, functionality, or style.

2. **Coding Standards**: Both authors and reviewers should ensure the code adheres to established coding standards, including formatting, naming conventions, and best practices as outlined in official Golang guidelines.

3. **Testing**: Reviewers should ensure that the submitted code is accompanied by adequate tests, covering both happy paths and error paths. Check the readability, descriptiveness, and structure of tests.

4. **Error Handling**: Pay special attention to error handling within the code. Ensure that errors are not ignored, logged appropriately, and presented to the user in a user-friendly format when applicable.

5. **Performance**: Review the code for any potential performance issues, such as inefficient loops, unnecessary allocations, or misuse of concurrency.

6. **Dependency Management**: Ensure that any new dependencies are necessary, appropriately versioned, and have been vetted for performance and security.

> 7. **Security Practices**: Scrutinize the code for any security vulnerabilities. Validate input, avoid security anti-patterns, and utilize secure coding practices.

Reviewer: Is this a clue for the reviewer? seems more like for the dev/author

Author: Replaced with point that is targeted to reviewers

- [ ] Do you understand the code and it's purpose?

This checklist serves as a guide to both authors and reviewers to ensure a thorough and effective code review process.

## 3. Contributing

### 3.1 Pull Requests && Issues

We have separate templates for Pull Requests and Issues. Please use them when creating a new PR or Issue.

### 3.2 Conventional Commits & Pull Requests Naming

#### 3.2.1 Overview

In an effort to maintain clarity and coherence in our commit history, we are adopting the Conventional Commits style for all commit messages across our repositories. This uniform format not only enhances the readability of our commit history but also facilitates automated tools in generating changelogs and extracting valuable information effectively.

#### 3.2.2 Structure

Conventional Commits follow a structured format: `type(scope): description`, where:

- `type`: Represents the nature of the commit (e.g., feat, fix, chore).
- `scope`: Denotes the relevant module or issue.
- `description`: Provides a brief explanation of the change.

When introducing breaking changes, an `!` should be appended after the `type/scope`: `feat(#123)!: introduce a breaking change`.

> 

Reviewer 2: Since we want to introduce automatic versioning based on conventional commits, maybe we can explain that here as well, what commit will trigger what version update:

`feat(#123)!: introduce breaking change` - `1.0.0` -> `2.0.0`
`feat(#124): introduce new feature` - `2.0.0` -> `2.1.0`
`fix(#125): fix a bug` - `2.1.0` -> `2.1.1`

Or, if we don't have automatic version updating, we can explain with this example to what version should someone bump when pushing a feature.

Reviewer: Good point!

Author: Added automatic versioning section :)

- `docs`: Designated for commits involving documentation changes, such as updating README files or adding comments.
- `style`: Applied to commits focusing on code style and formatting, without altering the code's functionality.
- `refactor`: Used for code changes that neither introduce new features nor fix bugs, but improve the code structure or design.
- `test`: Assigned to commits pertaining to the addition, modification, or refactoring of tests.
- `chore`: For changes related to build processes, local development, or other maintenance tasks.
- `perf`: Employed when enhancing the performance of the codebase.
- `revert`: Marked for commits that revert a previous change.
- `ci`: Applied to changes concerning the Continuous Integration (CI) configuration or scripts.
- `deps`: Used when updating or modifying dependencies.

#### 3.2.4 Scope

We have standardized the use of JIRA/GitHub issue numbers as the `scope` in commits within our team. This practice aids in easily tracing the origin of changes.

In the absence of an existing issue for your changes, please create one in the client’s JIRA system. If the change is not client-related, establish a GitHub issue in the repository.

> 

Reviewer 2: What if someone doesn't have access to our JIRA? I think we can say here that in that case the Scope can also be a regular scope like `feat(User): create new addUser endpoint`, or `feat(api): create new addUser endpoint`.

And maybe, just maybe, we could say that if the scope is not easy to explain in one word, there can be no scope, but it's an exception and then it needs a good commit message. That is okay with the rules from https://www.conventionalcommits.org/en/v1.0.0/ .

WDYT?

Reviewer: Any contributor who doesn't have access to JIRA should use GitHub issue what is mentioned above

#### 3.2.5 Further Reading

Additional information and guidelines on Conventional Commits can be found [here](https://www.conventionalcommits.org/en/v1.0.0/).

#### 3.2.6 Examples

##### Commit message with scope

Good example:

```bash

> feat: add possibility to create a new user by admin

Reviewer: Actually there is no scope in the example

Author: fixed

```

Bad example:

```bash
debugo feature - checkpoint full work
```

##### Pull request title with a scope and task number

> feat(#123): add new feature

### 3.3 Branching

#### Choosing branch names

- Choose consistent naming conventions. Common practices include:
  - `feature/feature-name`
  - `bugfix/issue-or-bug-name`
  - `hotfix/hotfix-name`
  - `chore/task-name`
  - `refactor/refactor-name`

#### Descriptiveness

- Branch names should be descriptive and represent the task/feature at hand.
- Use hyphens to separate words for readability, e.g., `feature/add-login-button`.

#### Include Issue Number

- If applicable, include the issue number in the branch name for easy tracking, e.g., `feature/123-add-login-button`.

#### Deleting Branches After Merging

#### Remove Remote Branches

- Once a PR has been merged, delete the remote branch to keep the repository clean.
- GitHub provides a button to delete the branch once the PR is merged.

#### Recommendation: Clean Local Branches

> - Regularly prune local branches that have been deleted remotely with `git fetch -p && git branch -vv | grep 'origin/.*: gone]' | awk '{print $1}' | xargs git branch -d`.

Reviewer: I like it!

- `test`: Assigned to commits pertaining to the addition, modification, or refactoring of tests.
- `chore`: For changes related to build processes, local development, or other maintenance tasks.
- `perf`: Employed when enhancing the performance of the codebase.
- `revert`: Marked for commits that revert a previous change.
- `ci`: Applied to changes concerning the Continuous Integration (CI) configuration or scripts.
- `deps`: Used when updating or modifying dependencies.

#### 3.2.4 Conventional Commits - Automatic Versioning

In our repositories, we use Conventional Commits to automatically generate the version number for our releases.

It works like this:

`fix: which represents bug fixes, and correlates to a SemVer patch.`
`feat: which represents a new feature, and correlates to a SemVer minor.`
`feat!:, or fix!:, refactor!:, etc., which represent a breaking change (indicated by the !) and will result in a SemVer major.`

Given a version number MAJOR.MINOR.PATCH, increment the:

MAJOR version when you make incompatible API changes
MINOR version when you add functionality in a backward compatible manner
PATCH version when you make backward compatible bug fixes

> Additional labels for pre-release and build metadata are available as extensions to the MAJOR.MINOR.PATCH format.

Reviewer 2: Very well explained, although I really liked my examples with `commit`: `version` -> `version` :)

If you decide to not include them in addition to your explanation, I completely understand, but I will be sad 😢 

Other than that, great job, approve from me ✅

Author: added Kuba - thought I added it earlier, sorry :(
